### PR TITLE
OCPBUGS-18783: e2e: perfprof: enhance the scheduling domain tests

### PR DESCRIPTION
**Pull request by ffromani**
Add simple parser for `/proc/schedstat`, to have a good foundation for load balancing disabling tests.
Use it into the scheduling domains tests to make the test more precise and (hopefully) the failure conditions clearer.
Finally, builds on the aforementioned to tight up the tests.

### CODE REVIEW

File: test/e2e/performanceprofile/functests/utils/pods/pods.go [hunk 1]
```diff
@@ -220,6 +221,19 @@ func DumpResourceRequirements(pod *corev1.Pod) string {
 	return sb.String()
 }
 
+func GetPodsOnNode(nodeName string) ([]corev1.Pod, error) {
+	listOptions := &client.ListOptions{
+		FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nodeName}),
+	}
+
+	pods := corev1.PodList{}
+	err := testclient.Client.List(context.TODO(), &pods, listOptions)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/utils/pods/pods.go:230]: Let's pass the context from outside the function

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 2]
```diff
@@ -263,7 +263,23 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+					testlog.Warningf("cannot list pods on %q: %v", workerRTNode.Name, err)
+				} else {
+					testlog.Infof("pods on %q BEGIN", workerRTNode.Name)
+					for _, pod := range pods {
+						testlog.Infof("- %s/%s %s annotations=%#v", pod.Namespace, pod.Name, pod.UID, pod.Annotations)
+					}
+					testlog.Infof("pods on %q END", workerRTNode.Name)
+				}
+
+				// TODO: should this be a skip?
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:279]: The base assumption is that no pods with lb disabled are in the cluster in the beginning of the test, hence we should Fail here.

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 3]
```diff
@@ -263,7 +263,23 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+				Fail(fmt.Sprintf("the test expects all CPUs within a scheduling domain, got %v instead", defaultCpuNotInSchedulingDomain))
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:280]: We can use `Expect` here since it's standard `ginkgo` assertion.

**Review by Tal-or — COMMENTED** [review]
Looking good, left couple of notes.

File: pkg/performanceprofile/utils/schedstat/schedstat.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		ret = append(ret, cpu)
+	}
+	return ret
+}
+
+func (inf Info) GetDomains(cpu string) []string {
+	return inf.cpuDomains[cpu]
+}
+
+func Parse() (Info, error) {
```
[the remark this answers is not in the document]
Comment (ffromani) [reply pkg/performanceprofile/utils/schedstat/schedstat.go]: That's initial draft never cleaned up. Done now.

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 3 above)
Comment (ffromani) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: Agreed, changed to Expect

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 2 above)
Comment (ffromani) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: addressed below. Thanks!

**Comment by ffromani** [comment]
/retest

this change can't break unrelated lanes. One last try before we start to deep dive nto actual failures

**Comment by ffromani** [comment]
/retest

infra issue

**Comment by ffromani** [comment]
/retest

apiserver failure

File: pkg/performanceprofile/utils/schedstat/schedstat_test.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		},
+	}
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ParseData(strings.NewReader(tt.schedStatData))
+			if err != nil {
+				t.Errorf("failure: unexpected error %v", err)
+			}
+			gotCPUs := got.GetCPUs()
+			if !reflect.DeepEqual(gotCPUs, tt.expectedCPUs) {
```
Comment (yanirq) [inline pkg/performanceprofile/utils/schedstat/schedstat_test.go:42]: does the order of the cpus matter here ?  will expectedCPUs be always sorted as well? meaning, do we need to enforce that before the comparison?

File: pkg/performanceprofile/utils/schedstat/schedstat.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+func (inf Info) GetDomainsByID(cpuID int) ([]string, bool) {
+	return inf.GetDomains(fmt.Sprintf("cpu%d", cpuID))
+}
+
+func (inf Info) GetDomains(cpu string) ([]string, bool) {
+	doms, ok := inf.cpuDomains[cpu]
+	return doms, ok
+}
+
+func ParseData(r io.Reader) (Info, error) {
```
Comment (yanirq) [inline pkg/performanceprofile/utils/schedstat/schedstat.go:54]: I wonder if we should add a comment on this function expected input/output for future use (or code is enough by itself)

File: pkg/performanceprofile/utils/schedstat/schedstat_test.go (hunk 5 above)
Comment (ffromani) [reply pkg/performanceprofile/utils/schedstat/schedstat_test.go:42]: yup, `GetCPUs` guarantees to return a sorted set

Comment (ffromani) [reply pkg/performanceprofile/utils/schedstat/schedstat_test.go:42]: OTOH I should sort `expectedCPUs` explicitely

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 7]
```diff
@@ -230,40 +232,73 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
 	When("pod runs with the CPU load balancing runtime class", func() {
 		var smtLevel int
+		var allTestpods map[types.UID]*corev1.Pod
 		var testpod *corev1.Pod
 
 		// getCPUswithLoadBalanceDisabled Return cpus which are not in any scheduling domain
 		getCPUswithLoadBalanceDisabled := func() ([]string, error) {
 			cmd := []string{"/bin/bash", "-c", "cat /proc/schedstat"}
-			schedstat, err := nodes.ExecCommandOnNode(cmd, workerRTNode)
+			schedstatData, err := nodes.ExecCommandOnNode(cmd, workerRTNode)
```
Comment (yanirq) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:241]: Maybe if we are changing this area already we can consider extarcting `getCPUswithLoadBalanceDisabled` function for better readability if an error occurs. Not a must.

Comment (ffromani) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:241]: good idea, on it

**Comment by yanirq** [comment]
looks good to me overall.  do we need all the separate commits here or can we squash them all into one ?

**Comment by ffromani** [comment]
> looks good to me overall. do we need all the separate commits here or can we squash them all into one ?

I don't see a reason to squash them, but if we prefer this way I don't mind too much

**Comment by yanirq** [comment]
/lgtm
/hold
remove hold once all required peers have a review

**Comment by ffromani** [comment]
/retest

```
received unexpected HTTP status: 504 Gateway Time-out
```

**Comment by ffromani** [comment]
/retest

more timeouts, more 504s

**Comment by ffromani** [comment]
/retest

sigh

**Pull request merged**